### PR TITLE
Guarded transaction

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: data/transaction/constants.go [hunk 1]
```diff
@@ -16,3 +16,7 @@ const (
 	// TxTypeInvalid represents the identifier for an invalid transaction
 	TxTypeInvalid TxType = "invalid"
 )
+
+// MaskSignedWithHash this mask used to verify if LSB from last byte from field options from transaction is set
+const MaskSignedWithHash     = uint32(1)
+const MaskGuardedTransaction = uint32(1) << 1
```
Comment (iulianpascalau) [inline data/transaction/constants.go:22]: empty line + add comment for MaskGuardedTransaction

File: data/interface.go [hunk 2]
```diff
@@ -273,6 +273,26 @@ type TransactionHandler interface {
...
 }
 
+// Encoder represents a byte slice to string encoder
+type Encoder interface {
+	Encode(buff []byte) string
+	IsInterfaceNil() bool
+}
+
+// Marshaller is able to encode an object to its byte slice representation
+type Marshaller interface {
```
Comment (iulianpascalau) [inline data/interface.go:283]: :+1: :+1: :+1:

File: data/transaction/transaction.go [hunk 3]
```diff
@@ -57,7 +57,7 @@ func TrimSliceHandler(in []data.TransactionHandler) []data.TransactionHandler {
 }
 
 // GetDataForSigning returns the serialized transaction having an empty signature field
-func (tx *Transaction) GetDataForSigning(encoder Encoder, marshalizer Marshalizer) ([]byte, error) {
+func (tx *Transaction) GetDataForSigning(encoder data.Encoder, marshalizer data.Marshaller) ([]byte, error) {
```
Comment (iulianpascalau) [inline data/transaction/transaction.go:60]: maybe rename marshalizer param also?

File: data/transaction/transaction.go [hunk 4]
```diff
@@ -80,9 +80,23 @@ func (tx *Transaction) GetDataForSigning(encoder Encoder, marshalizer Marshalize
...
 
+	if len(tx.GuardianAddr) > 0 {
+		ftx.GuardianAddr = encoder.Encode(tx.GuardianAddr)
+	}
+
 	return marshalizer.Marshal(ftx)
 }
 
+// HasOptionGuardianSet returns true if the guarded transaction option is set
+func (tx *Transaction) HasOptionGuardianSet() bool {
```
Comment (iulianpascalau) [inline data/transaction/transaction.go:91]: some tests for L91-L99 would have been great

File: data/transaction/constants.go (hunk 1 above)
Comment (AdoAdoAdo) [reply data/transaction/constants.go:22]: done

File: data/transaction/transaction.go (hunk 3 above)
Comment (AdoAdoAdo) [reply data/transaction/transaction.go:60]: done

File: data/transaction/transaction.go (hunk 4 above)
Comment (AdoAdoAdo) [reply data/transaction/transaction.go:91]: done

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

File: data/transaction/transaction.go [hunk 5]
```diff
@@ -80,7 +80,21 @@ func (tx *Transaction) GetDataForSigning(encoder Encoder, marshalizer Marshalize
 		Options:          tx.Options,
 	}
 
-	return marshalizer.Marshal(ftx)
+	if len(tx.GuardianAddr) > 0 {
+		ftx.GuardianAddr = encoder.Encode(tx.GuardianAddr)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply data/transaction/transaction.go:84]: it's fine like this

**Pull request merged**